website/pages/docs/secrets/databases/custom.mdx

```diff
@@ -12,81 +12,166 @@ description: |-
+
+	// UpdateUser updates an existing user within the database.
+	UpdateUser(ctx context.Context, req UpdateUserRequest) (UpdateUserResponse, error)
+
+	// DeleteUser from the database. This should not error if the user didn't
+	// exist prior to this call.
+	DeleteUser(ctx context.Context, req DeleteUserRequest) (DeleteUserResponse, error)
+
+	// Type returns the Name for the particular database backend implementation.
+	// This type name is usually set as a constant within the database backend
+	// implementation, e.g. "mysql" for the MySQL database backend. This is used
+	// for things like metrics and logging. No behavior is switched on this.
 	Type() (string, error)
-	CreateUser(ctx context.Context, statements Statements, usernameConfig UsernameConfig, expiration time.Time) (username string, password string, err error)
-	RenewUser(ctx context.Context, statements Statements, username string, expiration time.Time) error
-	RevokeUser(ctx context.Context, statements Statements, username string) error
-	RotateRootCredentials(ctx context.Context, statements []string) (config map[string]interface{}, err error)
-	Init(ctx context.Context, config map[string]interface{}, verifyConnection bool) (saveConfig map[string]interface{}, err error)
+
+	// Close attempts to close the underlying database connection that was
+	// established by the backend.
 	Close() error
 }
 ```
 
-You'll notice the first parameter to a number of those functions is a
-`Statements` struct. This struct is used to pass the Role's configured
-statements to the plugin on function call. The struct is defined as:
+Each of the request and response objects can also be found in `sdk/database/dbplugin/v5/database.go`.
 
-```go
-type Statements struct {
-	Creation   []string
-	Revocation []string
-	Rollback   []string
-	Renewal    []string
-}
-```
+In each of the requests, you will see a `Statements` object. This object represents the set of
+commands to run for that particular operation. For the `NewUser` function, this is a set of
+commands to create the user (and often set permissions for that user). These statements are
+from the following fields in the API:
 
-It is up to your plugin to replace the `{{name}}`, `{{password}}`, and
-`{{expiration}}` in these statements with the proper values.
+| API Argument               | Request Object                                     |
+| -------------------------- | -------------------------------------------------- |
+| `creation_statements`      | `NewUserRequest.Statements.Commands`               |
+| `revocation_statements`    | `DeleteUserRequest.Statements.Commands`            |
+| `rollback_statements`      | `NewUserRequest.RollbackStatements.Commands`       |
+| `renew_statements`         | `UpdateUserRequest.Expiration.Statements.Commands` |
+| `rotation_statements`      | `UpdateUserRequest.Password.Statements.Commands`   |
+| `root_rotation_statements` | `UpdateUserRequest.Password.Statements.Commands`   |
```

Comment:
Do we state anywhere what happens if you set both `rotation_statements` and `root_rotation_statements`?

Separately to this PR, perhaps we should mark `root_rotation_statements` as deprecated?

Reply:
`root_rotation_statements` is not deprecated. The DB engine uses it when the `/database/rotate-root/:name` endpoint is called, whereas `rotation_statements` is used during a static credential rotation. The DB engine selects which one based on what action it is performing.

Reply:
Ah I read too much into how the config was actually stored from this table. Makes sense, thanks for clearing that up for me.